### PR TITLE
[IMP] repair,stock: allow creation/link of repairs from returns

### PR DESCRIPTION
**Pull request by ticodoo**
This feature supports the use case of:
- a customer receives a delivery from you.
- there is a defective product in the delivery
- they schedule a repair (RO) + return (picking) for the product

We want to ease the creation/management of the repair order associated
with the return, therefore the following features are added:

- Create links between ROs and return pickings (smart button in return
  of all repairs associated with it + field in RO)
- Option (requires active setting) to create RO(s) directly from a
  return picking

Implementation notes:
- Because return picking type doesn't have its own "Type of Operation"
  (it is 'code'='incoming'), we default to only incoming picking types
  that have at least 1 other picking type it is the
  'return_picking_type_id' of.
- Originally a Wizard for choosing which/how many products should have
  ROs created for them was designed (i.e. to match return wizard + to
  multi-create ROs), but this was determined to be too complicated due
  to SN/lot selection. Instead simple approach of creating a single RO
  each time button is pushed was implemented to match the helpdesk >
  repair button workflow.
- ROs are purposely allowed for 'draft' and 'cancelled' returns/moves to
  support different types of workflows (e.g. RO is confirmed before
  return is)

Task: 2732517

### CODE REVIEW

File: addons/repair/models/repair.py [hunk 1]
```diff
@@ -111,6 +111,13 @@ class Repair(models.Model):
         readonly=True, required=True, index=True,
         default=lambda self: self.env.company)
     sale_order_id = fields.Many2one('sale.order', 'Sale Order', copy=False, help="Sale Order from which the product to be repaired comes from.")
+    picking_id = fields.Many2one(
+        'stock.picking', 'Return',
```
Comment (amoyaux) [inline addons/repair/models/repair.py:115]: check_company=True,

File: addons/repair/models/stock_picking.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        "Technical field to identify default return picking type for a warehouse.", store=True,
+        compute='_compute_can_repair')
+
+    @api.depends('warehouse_id.return_type_id')
+    def _compute_can_repair(self):
+        self.is_return_type = False
+        for picking_type in self:
+            picking_type.is_return_type = picking_type == picking_type.warehouse_id.return_type_id
+            if not picking_type.is_return_type:
+                picking_type.can_repair = False
```
Comment (amoyaux) [inline addons/repair/models/stock_picking.py:24]: I think it's better to check if the picking.type is a return_picking_type_id of another picking.type instead of checking on the warehouse. It will restrict the possible configuration

Comment (amoyaux) [reply addons/repair/models/stock_picking.py:24]: Maybe with a one2many field for the depends trigger

File: addons/repair/models/stock_picking.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    nbr_repairs = fields.Integer('Number of repairs linked to this picking', compute='_compute_nbr_repairs')
+
+    @api.depends('repair_ids')
+    def _compute_nbr_repairs(self):
+        for picking in self:
+            picking.nbr_repairs = len(picking.repair_ids)
+
+    def action_repair_return(self):
+        self.ensure_one()
+        ctx = self._context.copy()
```
Comment (amoyaux) [inline addons/repair/models/stock_picking.py:41]: self.env.context

File: addons/repair/models/stock_picking.py (hunk 2 above)
Comment (ticodoo) [reply addons/repair/models/stock_picking.py:24]: This is a tricky thing unfortunately. The feature is specifically only supposed to be for the "Returns" picking type. Since there's no `code=return` option, I originally was inferring the "Returns" from operation types with `code=incoming` + no `return_picking_type_id` to distinguish it from the default "Receipt" type. However the flexibility to create new alternative "Return" types wasn't desired + it also opened the possibility of configuring the types incorrectly and having a weird repair option show up when it shouldn't.

TLDR version: I could make it check for whether a picking type is a `return_picking_type_id` + `code=incoming`, but this is a flexibility that isn't needed/could result in the "repair" option showing up when not desired.

File: addons/repair/models/repair.py (hunk 1 above)
Comment (ticodoo) [reply addons/repair/models/repair.py:115]: We can link repair orders to SOs with another company_id :thinking: do we not want to do the same for repairs? Or should I change the SO to check_company as well?.

File: addons/repair/models/stock_picking.py (hunk 3 above)
Comment (ticodoo) [reply addons/repair/models/stock_picking.py]: Done

File: addons/repair/models/repair.py (hunk 1 above)
Comment (ticodoo) [reply addons/repair/models/repair.py]: Done, also added/fixed for SOs.

File: addons/repair/models/stock_picking.py (hunk 2 above)
Comment (ticodoo) [reply addons/repair/models/stock_picking.py]: Ignore previous comment. This has been changed to only `picking.type.code == 'incoming'` + is a return_picking_type_id of another picking type as suggested.

**Comment by amoyaux** [comment]
robodoo r+ rebase-ff
